client/deposit_ethereum.go

```diff
@@ -0,0 +1,467 @@
+		// mayhul
+		chainID, err := c.ethClient.ChainID(context.Background())
+		if err != nil {
+			return nil, err
+		}
+
+		signedTx, err := types.SignTx(tx, types.NewLondonSigner(chainID), privateKey)
+		if err != nil {
+			return nil, err
+		}
+
+		return signedTx, nil
+	}
+
+	if currency == ETH {
+		opt, err := getTransactionOpt(ctx, c.ethClient, ethPrivateKey, signerFn, amount, blockchainDecimal)
+		if err != nil {
+			return CryptoDepositResponse{}, err
+		}
+
+		transaction, err = c.starkexContract.DepositEth(opt, starkPublicKeyBigInt, starkAssetIDBigInt, vaultIDBigInt)
+		if err != nil {
+			return CryptoDepositResponse{}, err
+		}
+
+	} else {
+		opt, err := getTransactionOpt(ctx, c.ethClient, ethPrivateKey, signerFn, 0, decimal)
+		if err != nil {
+			return CryptoDepositResponse{}, err
+		}
+		opt.GasLimit = 100000
+
+		if !c.ethereumNetworkAllowanceSet {
+			err = setAllowance(ctx, c.ethClient, coinStatus.TokenContract, opt, c.starkexContractAddress, blockchainDecimal, c.ethereumNetworkAllowance)
+			if err != nil {
+				return CryptoDepositResponse{}, err
+			}
+
+			c.ethereumNetworkAllowanceSet = true
+		}
+
+		allowance, err := getAllowance(c.ethClient, ethAddress, coinStatus.TokenContract, blockchainDecimal, c.starkexContractAddress)
+		if err != nil {
+			return CryptoDepositResponse{}, err
+		}
+
+		if allowance < amount {
+			return CryptoDepositResponse{}, fmt.Errorf("current allowance is %v. call SetAllowance function to change the allowance amount", allowance)
+		}
+
+		transaction, err = c.starkexContract.DepositERC20(opt, starkPublicKeyBigInt, starkAssetIDBigInt, vaultIDBigInt, quantizedAmount)
+		if err != nil {
+			return CryptoDepositResponse{}, err
+		}
+	}
+
+	var starkAssetId string
+	if coinStatus.StarkAssetID[0:3] == "0x0" {
+		starkAssetId = "0x" + coinStatus.StarkAssetID[3:]
+	} else {
+		starkAssetId = coinStatus.StarkAssetID
+	}
+
+	resp, err := c.cryptoDepositStart(ctx, CryptoDepositRequest{
+		Amount:                 quantizedAmount.String(),
+		StarkAssetID:           starkAssetId,
+		StarkPublicKey:         starkPublicKey,
+		DepositBlockchainHash:  transaction.Hash().Hex(),
+		DepositBlockchainNonce: fmt.Sprintf("%v", transaction.Nonce()),
+		VaultID:                vaultID,
+	})
+	if err != nil {
+		return CryptoDepositResponse{}, err
+	}
+
+	resp.Payload = transaction.Hash().Hex()
+	return resp, nil
+}
+
+type StarkexContract interface {
+	DepositEth(opts *bind.TransactOpts, starkKey *big.Int, assetType *big.Int, vaultId *big.Int) (*types.Transaction, error)
+	DepositERC20(opts *bind.TransactOpts, starkKey *big.Int, assetType *big.Int, vaultId *big.Int, quantizedAmount *big.Int) (*types.Transaction, error)
+}
+
+// Note: This function has to be called before using any other deposit function
+func (c *Client) DepositFromEthereumNetworkInit(ctx context.Context, rpcURL string) error {
+	ethClient, err := ethclient.Dial(rpcURL)
+	if err != nil {
+		return err
+	}
+	c.ethClient = ethClient
```

Comment:
instead of adding an extra init function, following can be done in deposit function:
- check if the required variables are initiated
- if not, initiate them
- continue with normal deposit flow

Reply:
Ok